wandb/sdk/launch/docker.py

```diff
@@ -45,53 +40,273 @@ def validate_docker_installation() -> None:
+                "python3-pip",
+                "python3-setuptools",
+            ]
+        base_setup = CUDA_SETUP_TEMPLATE.format(
+            cuda_base_image="nvidia/cuda:{}-runtime".format(cuda_version),
+            python_packages=" \\\n".join(python_packages),
+            py_version=py_version,
+        )
+    else:
+        python_packages = [
+            "python3-dev" if py_major == "3" else "python-dev",
+            "gcc",
+        ]  # gcc required for python < 3.7 for some reason
+        base_setup = PYTHON_SETUP_TEMPLATE.format(py_base_image=python_base_image)
+    return base_setup
+
+
+def get_env_vars_section(launch_project: LaunchProject, api: Api, workdir: str) -> str:
+    """Fill in wandb-specific environment variables"""
+
+    if _is_wandb_local_uri(api.settings("base_url")) and sys.platform == "darwin":
+        _, _, port = _, _, port = api.settings("base_url").split(":")
+        base_url = "http://host.docker.internal:{}".format(port)
+    elif _is_wandb_dev_uri(api.settings("base_url")):
+        base_url = "http://host.docker.internal:9002"
+    else:
+        base_url = api.settings("base_url")
+    return "\n".join(
+        [
+            f"ENV WANDB_BASE_URL={base_url}",
+            f"ENV WANDB_API_KEY={api.api_key}",
+            f"ENV WANDB_PROJECT={launch_project.target_project}",
+            f"ENV WANDB_ENTITY={launch_project.target_entity}",
+            f"ENV WANDB_LAUNCH={True}",
+            f"ENV WANDB_LAUNCH_CONFIG_PATH={os.path.join(workdir, DEFAULT_LAUNCH_METADATA_PATH)}",
+            f"ENV WANDB_RUN_ID={launch_project.run_id or None}",
+            f"ENV WANDB_DOCKER={launch_project.docker_image}",
+        ]
     )
-    assert isinstance(path, str)
-    _logger.info("Running repo2docker...")
-    cmd: Sequence[str] = [
-        "jupyter-repo2docker",
-        "--no-run",
-        "--image-name={}".format(launch_project.base_image),
-        "--user-id={}".format(launch_project.docker_user_id),
-        path,
-        '"{}"'.format(entry_cmd),
-    ]
-
-    _logger.info(
-        "Generating docker base image from git repo or finding image if it already exists..."
+
+
+def get_requirements_section(launch_project: LaunchProject) -> str:
+    if launch_project.deps_type == "pip":
+        requirements_files = ["src/requirements.txt"]
+        pip_install_line = "pip install -r requirements.txt"
+        if os.path.exists(
+            os.path.join(launch_project.project_dir, "requirements.frozen.txt")
+        ):
+            # if we have frozen requirements stored, copy those over and have them take precedence
+            requirements_files += ["src/requirements.frozen.txt", "_wandb_bootstrap.py"]
+            pip_install_line = (
+                _parse_existing_requirements(launch_project)
+                + "python _wandb_bootstrap.py"
+            )
+
+        if docker.is_buildx_installed():
+            prefix = "RUN --mount=type=cache,mode=0777,target=/root/.cache/pip"
```

Comment:
note that because we are doing a multistage build and the requirement installation takes place in the first stage, we can drop the uid/gid requirements from the buildkit configuration and do all installs as root. all installed deps are copied over in an env to the actual docker image provided to the user, which contains the setup for the user/uid/gid